### PR TITLE
Working tests for includes, sources and samples. Minor bugfixes

### PR DESCRIPTION
**Pull request by BarkingBad**

### CODE REVIEW

**Comment by BarkingBad** [comment]
Do not merge

File: plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt [hunk 1]
```diff
@@ -55,19 +56,35 @@ abstract class SamplesTransformer(val context: DokkaContext) : PageTransformer {
...
+            is ContentDRILink -> copy(children.map { it.bfs(fqName, node) })
+            is ContentResolvedLink -> copy(children.map { it.bfs(fqName, node) })
+            is ContentEmbeddedResource -> copy(children.map { it.bfs(fqName, node) })
+            is ContentTable -> copy(children = children.map { it.bfs(fqName, node) as ContentGroup })
+            is ContentList -> copy(children.map { it.bfs(fqName, node) })
+            is ContentGroup -> copy(children.map { it.bfs(fqName, node) })
+            is PlatformHintedContent -> copy(inner.bfs(fqName, node))
+            is ContentText -> if (text == fqName) node else this
+            else -> this
+        }
```
Comment (BarkingBad) [inline plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt:87]: Maybe most of these tags could point to `this`, but I'm not sure when one of these would be wrapper for some content in description

Comment (BarkingBad) [reply plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt:87]: Also it has a boilerplate, but I need smartcast for `copy` function

File: plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt [hunk 2]
```diff
@@ -55,19 +56,35 @@ abstract class SamplesTransformer(val context: DokkaContext) : PageTransformer {
...
             return this.also { context.logger.warn("Cannot find PsiElement corresponding to $fqName") }
         val imports = processImports(psiElement) // TODO: Process somehow imports. Maybe just attach them at the top of each body
         val body = processBody(psiElement)
-        val node = platformHintedContentCode(platform, contentPage.dri, body, "kotlin")
-        return this.safeAs<ContentGroup>()?.run { copy(
-            children = children.indexOfFirst { contentNode ->
-                contentNode.safeAs<ContentHeader>()?.children?.firstOrNull()?.safeAs<ContentText>()?.text == "Sample"
-            }.takeIf { it != -1 }?.let { children.apply { this.safeAs<MutableList<ContentNode>>()?.add(it+1, node) } } ?: children.also { context.logger.warn("Not found Sample block in ${contentPage.dri}")}
-        ) } ?: this.also { context.logger.warn("ContentPage ${contentPage.dri} cannot be cast to ContentGroup") }
+        val node = contentCode(contentPage.platforms(), contentPage.dri, body, "kotlin")
```
Comment (kamilok1965) [inline plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt:69]: You can at least inline `val body` or both of them

File: plugins/base/src/main/kotlin/translators/documentables/DefaultPageCreator.kt [hunk 3]
```diff
@@ -299,6 +299,23 @@ open class DefaultPageCreator(
             }
         }
 
+        fun DocumentableContentBuilder.contentForSamples() {
+            val samples = tags.withTypeNamed<Sample>()
+            if (samples.isNotEmpty()) {
+                platforms.forEach { platformData ->
+                    val content = samples.filter { it.value.isEmpty() || platformData in it.value }
+                    if (content.isNotEmpty()) {
+                        group(platformData = setOf(platformData)) {
```
Comment (kamilok1965) [inline plugins/base/src/main/kotlin/translators/documentables/DefaultPageCreator.kt:308]: I'd give the `ContentGroup` a `ContentKind.Sample` kind

File: plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt [hunk 4]
```diff
@@ -499,12 +499,15 @@ private class DokkaDescriptorVisitor(
...
             .map { enumEntryDescriptor(it, parent) }
 
-
-    private fun DeclarationDescriptor.resolveDescriptorData(platformData: PlatformData?): PlatformDependent<DocumentationNode> =
-        if (platformData != null) PlatformDependent.from(
-            platformData,
-            getDocumentation()
-        ) else PlatformDependent.expectFrom(getDocumentation())
+    private fun DeclarationDescriptor.resolveDescriptorData(platformData: PlatformData?): PlatformDependent<DocumentationNode> {
+        val documentation = getDocumentation()
```
Comment (kamilok1965) [inline plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:503]: `getDocumentation().let {}`, no need for this temp variable there IMHO

File: plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt [hunk 5]
```diff
@@ -2,83 +2,104 @@ package org.jetbrains.dokka.base.transformers.pages.sourcelinks
...
 import org.jetbrains.kotlin.descriptors.DeclarationDescriptorWithSource
-import org.jetbrains.kotlin.kdoc.psi.api.KDoc
-import org.jetbrains.kotlin.psi.KtDeclaration
 import org.jetbrains.kotlin.resolve.source.getPsi
 import org.jetbrains.kotlin.utils.addToStdlib.cast
 import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
-class SourceLinksTransformer(val context: DokkaContext) : PageTransformer {
+class SourceLinksTransformer(val context: DokkaContext, val builder: PageContentBuilder) : PageTransformer {
+    private val sourceLinks = getSourceLinks()
```
Comment (kamilok1965) [inline plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt:20]: Looks like it's used only in one place, I'd inline it then

File: plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt [hunk 6]
```diff
@@ -2,83 +2,104 @@ package org.jetbrains.dokka.base.transformers.pages.sourcelinks
...
-                sourceLinks.find { entry.value.path.contains(it.path) && it.platformData == entry.key }?.run {
-                    acc.modified(
-                        content = acc.content.addSource(
-                            entry.key,
-                            entry.value.cast<DescriptorDocumentableSource>().toLink(this)
-                        )
-                    )
-                } ?: acc
-            } ?: node
+    private fun resolveSources(documentable: WithExpectActual) = documentable.sources.map.entries
```
Comment (kamilok1965) [inline plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt:36]: I think you can remove the `.map.entries` part entirely

File: plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt (hunk 2 above)
Comment (BarkingBad) [reply plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt:69]: Is is likely that proper samples presentation requires further manipulations on imports/body. I think inlining won't change much.

File: plugins/base/src/main/kotlin/translators/documentables/DefaultPageCreator.kt (hunk 3 above)
Comment (BarkingBad) [reply plugins/base/src/main/kotlin/translators/documentables/DefaultPageCreator.kt:308]: Done

File: plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt (hunk 5 above)
Comment (BarkingBad) [reply plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt:20]: Done

File: plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt (hunk 4 above)
Comment (BarkingBad) [reply plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:503]: Done

File: plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt (hunk 6 above)
Comment (BarkingBad) [reply plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt:36]: Done

File: plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt [hunk 7]
```diff
@@ -55,19 +56,35 @@ abstract class SamplesTransformer(val context: DokkaContext) : PageTransformer {
         }
     }.toMap()
 
-    private fun ContentNode.addSample(contentPage: ContentPage, platform: PlatformData, fqName: String, analysis: Map<PlatformData, EnvironmentAndFacade>): ContentNode {
-        val facade = analysis[platform]?.facade ?:
-            return this.also { context.logger.warn("Cannot resolve facade for platform ${platform.name}")}
+    private fun ContentNode.addSample(contentPage: ContentPage, platform: PlatformData?, fqName: String, analysis: Map<PlatformData, EnvironmentAndFacade>): ContentNode {
+        val facade = if(platform == null) {
+            analysis.entries.find { it.key.platformType.name == "common" }?.value
```
[the remark this answers is not in the document]
Comment (BarkingBad) [reply plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt:61]: I think it won't work anymore after @sellophane changes with PlatformData

**Pull request merged**